### PR TITLE
Replace child process with u32 pid in Machine type

### PR DESCRIPTION
**Pull request by loony-bean**
This will allow creating instances of Machine without actually spinning up new VMs

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,22 +1,23 @@
 [package]
-name = "firec"
 description = "Rust API to interact with firecracker"
-version = "0.1.0"
 edition = "2021"
+name = "firec"
+version = "0.1.0"
```
Comment (zeenix) [inline Cargo.toml:5]: Unrelated change, please remove.

File: Cargo.toml [hunk 2]
```diff
@@ -1,22 +1,23 @@
...
-tokio = { version = "1.18.1", features = ["process", "net", "fs", "time"] }
+tokio = {version = "1.18.1", features = ["process", "net", "fs", "time"]}
 tracing = "0.1.34"
 users = "0.11.0"
-uuid = { version = "1.0.0", features = ["serde", "v4"] }
+uuid = {version = "1.0.0", features = ["serde", "v4"]}
 
 [dev-dependencies]
-tokio = { version = "1.18.1", features = ["rt", "macros"] }
+tokio = {version = "1.18.1", features = ["rt", "macros"]}
```
Comment (zeenix) [inline Cargo.toml:23]: lots of irrelevant changes here too. Please split into a separate commit.

File: src/machine.rs [hunk 3]
```diff
@@ -235,7 +236,8 @@ impl<'m> Machine<'m> {
     pub async fn stop(&mut self) -> Result<(), Error> {
         let vm_id = self.vm_id();
         trace!("{vm_id}: Killing VM...");
-        self.child.kill().await?;
+        let child = heim::process::get(self.pid.try_into().unwrap()).await?;
```
Comment (zeenix) [inline src/machine.rs:239]: Let's not `unwrap` the conversion error even if unlikely to happen. Let's just wrap [TryFromIntError](https://doc.rust-lang.org/nightly/std/num/struct.TryFromIntError.html) in our error.

File: src/machine.rs [hunk 4]
```diff
@@ -191,6 +191,7 @@ impl<'m> Machine<'m> {
         }
         trace!("{vm_id}: Running command: {:?}", cmd);
         let child = cmd.spawn()?;
+        let pid = child.id().expect("firecracker process exited early");
```
Comment (zeenix) [inline src/machine.rs:194]: Failure here shouldn't panic I think and doesn't seem too unlikely to happen. Maybe add a new Error variant for this case?

Comment (zeenix) [reply src/machine.rs:194]: We probably should also include the return value of the process, which we can get using `Child::wait` method if this is `None`.

File: Cargo.toml (hunk 1 above)
Comment (loony-bean) [reply Cargo.toml:5]: put into separate commit

File: Cargo.toml (hunk 2 above)
Comment (loony-bean) [reply Cargo.toml:23]: same here, put into separate commit

File: src/machine.rs (hunk 3 above)
Comment (loony-bean) [reply src/machine.rs]: done

File: src/machine.rs (hunk 4 above)
Comment (zeenix) [reply src/machine.rs]: > done

:+1: 

pro-tip: you can just click "Resolve conversation".

File: Cargo.toml (hunk 1 above)
Comment (zeenix) [reply Cargo.toml:5]: I don't get this part though. I'd expect the name to be the first thing listed, before description.

File: src/machine.rs [hunk 5]
```diff
@@ -190,7 +190,13 @@ impl<'m> Machine<'m> {
...
         }
         trace!("{vm_id}: Running command: {:?}", cmd);
-        let child = cmd.spawn()?;
+        let mut child = cmd.spawn()?;
+        let pid = if let Some(id) = child.id() {
+            Ok(id)
+        } else {
+            let exit_status = child.wait().await?;
+            Err(Error::ProcessExitedEarly { exit_status })
+        }?;
```
Comment (zeenix) [inline src/machine.rs:199]: I think a match would be better here:

```suggestion
        let pid = match child.id() {
            Some(id) => id,
            None => {
                let exit_status = child.wait().await?;
                return Err(Error::ProcessExitedEarly { exit_status });
        };
```

No?

File: Cargo.toml (hunk 1 above)
Comment (loony-bean) [reply Cargo.toml:5]: it's automatically sorted just as blocks of mod's or use's

Comment (loony-bean) [reply Cargo.toml:5]: oh that's not rust fmt work but my formatter for toml
so this is definitely debatable

Comment (zeenix) [reply Cargo.toml:5]: yeah, rustfmt only handles rust files (for now at least). Your toml formatter probably doesn't know about Cargo but works in a generic way.

I checked the Cargo.toml of top most downloaded crates on crates.io and they all have name and version as the first thing. I hope you can agree this makes sense.

**Review by zeenix — APPROVED** [review]
